### PR TITLE
 Training Pipeline Enhancements

### PR DESCRIPTION
**Pull request by sridhs21**
This PR enhances the training pipeline to address vanishing gradient problems that were preventing successful mixed precision training implementation. The primary focus is on implementing automatic mixed precision (AMP) support while making necessary architectural and pipeline improvements to ensure gradient stability. The changes include:

- Automatic Mixed Precision (AMP) training support:
  - Adds `--use-amp` and `--amp-dtype` flags to enable mixed precision training with both float16 and bfloat16 support. This can significantly reduce memory usage and training time on compatible GPUs while maintaining numerical stability through proper gradient scaling.

- Enhanced U-Net architecture with residual connections:
  - Replaces the basic U-Net with a residual block-based design that includes batch normalization and skip connections. This improves gradient flow and training stability while adding dropout layers to prevent overfitting, which was a common problem when implementing AMP.

- Improved training pipeline:
  - Implements patch-based training through `XPointPatchDataset` for better data augmentation, adds feature normalization for training stability, includes gradient clipping to prevent exploding gradients, and adds early stopping with patience to prevent overfitting. The data is also resampled when training, in that it is undersampled. Patches with no X-points are removed to match the number of patches with X-points present to create a balanced dataset.

- Enhanced optimization:
  - Switches from Adam to AdamW optimizer with weight decay for better generalization, adds cosine annealing learning rate scheduling, and improves checkpoint functionality to save/load all training state, including AMP scalers.

I have also updated the README file to include a description of what flags can be used.

### CODE REVIEW

**Review by cwsmith — CHANGES_REQUESTED** [review]
Looks good.  Thank you.  I appreciate the update to the README and the verbose PR description.  

Would you please remove the whitespace only changes?

Update: as discussed in the meeting, if changing the whitespace breaks python then please ignore the whitespace request

File: XPointMLTest.py [hunk 1]
```diff
@@ -274,159 +283,169 @@ def load(self, fnum):
...
+            "all": all_torch,      # Normalized for training
+            "mask": mask_torch,    # shape [1, Nx, Ny]
             "x": fields["coords"][0],
             "y": fields["coords"][1],
             "filenameBase": fields["fileName"],
             "params": dict(self.params)  # copy of the params for local plotting
         }
 
+class XPointPatchDataset(Dataset):
+    """On‑the‑fly square crops, balancing positive / background patches."""
```
Comment (cwsmith) [inline XPointMLTest.py:296]: Is this removing portions of the input frames that have no x-points?  If not, would you please explain a bit more what this is doing?

File: XPointMLTest.py [hunk 2]
```diff
@@ -857,53 +948,116 @@ def main():
...
+            torch.save(model.state_dict(), os.path.join(checkpoint_dir, "best_model.pt"))
+        else:
+            patience_counter += 1
+            
+        # Save checkpoint periodically
+        if (epoch+1) % args.checkPointFrequency == 0:
+            save_model_checkpoint(model, optimizer, train_loss, val_loss, epoch+1, checkpoint_dir, scaler, best_val_loss)
+        
+        # Early stopping
+        if patience_counter >= args.patience:
```
Comment (cwsmith) [inline XPointMLTest.py:1058]: This will stop training if the training/value loss isn't decreasing sufficiently?  If so, can you expand the doc string to indicate that.

File: XPointMLTest.py (hunk 1 above)
Comment (cwsmith) [reply XPointMLTest.py:296]: As discussed offline, this is removing subdomains that have no x-points to ensure that there are an equal number of subdomains that have x-points and those that don't.  Given that we have so few x-points, this is going to mostly be removing subdomains without x-points.

File: XPointMLTest.py (hunk 2 above)
Comment (cwsmith) [reply XPointMLTest.py:1058]: > This will stop training if the training/value loss isn't decreasing sufficiently

As discussed offline, yes.  'patience' is common a ML/AI training term/concept.
